packages/next/src/build/index.ts

```diff
@@ -567,16 +567,17 @@ export default async function build(
       const instrumentationHookEnabled = Boolean(
         config.experimental.instrumentationHook
       )
-      const rootPaths = (
-        await flatReaddir(rootDir, [
-          middlewareDetectionRegExp,
-          ...(instrumentationHookEnabled
-            ? [instrumentationHookDetectionRegExp]
-            : []),
-        ])
-      )
+
+      const includes = [
+        middlewareDetectionRegExp,
+        ...(instrumentationHookEnabled
+          ? [instrumentationHookDetectionRegExp]
+          : []),
+      ]
+
+      const rootPaths = (await getFilesInDir(rootDir))
+        .filter((file) => includes.some((include) => include.test(file)))
         .sort(sortByPageExts(config.pageExtensions))
-        .map((absoluteFile) => absoluteFile.replace(dir, ''))
```

Comment:
Small amount of irony in that previously `flatReaddir` returned an absolute path which is then iterated through again to remove. The updated function no longer returns an absolute path.

Reply:
😅 This was the cause of test failures since I didn't realise that that the end result would be `["/middleware.ts"]` and instead returned `["middleware.ts"]`. I've added it back in.